домашние задания 9-1 и 9-2 + исправления по 8 + 10-1 + 11-1-2-3-4 + 12-1-2-3

wrhmmstr opened the pull request.

Commit:
refactor: - комментарии и неиспользуемые импорты

Commit:
refactor: приём vararg декораторов на вход

Commit:
refactor: дефолтный размер + размер через конструктор

Добавил дефолтный размер страницы + размер через конструктор, в рамках замечений по практикам 8

Commit:
refactor: работа с единым массивом сообщений

Commit:
refactor: перебор варарга декораторов

Commit:
refactor: вынес логику порядка в отдельный класс

Commit:
refactor: вынес логику дублей в отдельный класс

Commit:
docs: дополнил javadoc

Commit:
feat: добавил валидацию входных параметров

добавил базовый общий класс с повторно используемым поведениемв для валиадцаии входных параметров в рамках практики 09-1: Борьба с дублированием с помощью наследования

Commit:
feat: переопределил методы класса Object

переопределил методы класса Object: toString(), equals() и hashCode, и добавил тестовые примеры в рамках практики 09-2: Встраивание своего дизайна в общие договоренности системной библиотеки Java

Commit:
feat: добавлена обработка исключений

добавлена обработка исключений в рамках практики 10-1: Реализация альтернативных сценариев поведения системы с помощью механизма exceptions

Commit:
refactor: уникальный массив без пустых элементов

Commit:
fix: проверка аргументов и тестовый пример

Commit:
feat: возможность выполнить 2 кейса с исключениями

Вынес тесты в отдельный класс, для выполнения 2 и более кейсов с исключениями в рамках 10-1: Реализация альтернативных сценариев поведения системы с помощью механизма exceptions

Commit:
feat: добавил сохранение в хэшмап

добавил сохранение декорированного сообщения, с присвоением UUID, в хэшмап, в рамках практики 11-1: Рефакторинг stateless системы в stateful учетную – добавление операции Create

Commit:
fix: исправил создание decoratedMessage

Commit:
feat: добавил операции чтения из коллекции

добавил операции чтения по ключу, уровню важности, всех сообщений, в рамках практик 11-2: Добавление операции Read по первичному ключу
11-3: Добавление операции Read всех сообщений, 11-4 [optional]: Добавление операции Read сообщений по уровню значимости

Commit:
fix: убрал логику невыполнения кода из 7 практики

Commit:
refactor: тесты под фреймворк JUnit 5

Изменил тесты в рамках практики 12-1: Рефакторинг полуавтоматических тестов в полноценные под фреймворк JUnit5

dmitrii-grog commented on src/main/java/com/tcs/edu/ExceptionLogger.java:
```diff
@@ -0,0 +1,23 @@
+package com.tcs.edu;
+
+public class ExceptionLogger extends RuntimeException {
```
Название сбивает столку, из названия я ожидаю что этот класс будет логировать исключения. Хотя по логике это класс исключения возникшего во время попытки залогировать, поэтому название и  "LogException" в задании.
Еще, из задачи, исключение должно быть проверяемым т.е. наследоваться от класса Exception

wrhmmstr replied:
Исправлено

Commit:
fix: проверяемое исключение + имя класса

замечания по 10 практике

Commit:
fix: добавил счетчик в конструктор

fix: добавил счетчик в конструктор, чтобы для каждого вызова счётчик был уникальный

Commit:
fix: исправил обработку пустого сообщения

Commit:
fix: добавил обработку кастомного исключения

Commit:
refactor: изменил тесты AssertJ + Nested Tests

refactor: изменил тесты AssertJ + Nested Tests в соответствии с практиками #12-2 [optional]: Использование комбинаторов утверждений и #12-3 [optional]: Использование тестовых наборов (suites)

dmitrii-grog commented on src/test/java/com/tcs/edu/HashMapMessageRepositoryTests.java:
```diff
@@ -0,0 +1,153 @@
+
+            //When
+            uuid = sut.create(message1);
+            System.out.println(uuid);
```
В тесте не должно в итоге остаться принтов, это только для отладки

wrhmmstr replied:
Исправлено

dmitrii-grog commented on src/test/java/com/tcs/edu/HashMapMessageRepositoryTests.java:
```diff
@@ -0,0 +1,153 @@
+    public void SetUp() {
+        sut = new HashMapMessageRepository();
+        currentMessage = null;
+        uuid = null;
```
Нет смысла выносить в поля класса переменные которые реально инициализируются локально уже в самом тесте

dmitrii-grog commented on src/test/java/com/tcs/edu/HashMapMessageRepositoryTests.java:
```diff
@@ -0,0 +1,153 @@
+            //Then
+            assertThat(currentMessage).isNotNull()
+                    .extracting("level", "message", "id")
+                    .contains(currentMessage.getLevel(), currentMessage.getMessage(), currentMessage.getId())
```
Странный способ проверить поля проверяемого объекта его же полями, в итоге это проверка не имеет смысла

wrhmmstr replied:
Исправлено

dmitrii-grog commented on src/main/java/com/tcs/edu/service/ValidatingService.java:
```diff
@@ -0,0 +1,42 @@
+        if (message.getMessage() == null) throw new IllegalArgumentException("Message message.getMessage() value is: " + message.getMessage() + ".");
+        if (message.getMessage().isEmpty()) throw new IllegalArgumentException("Message message.getMessage() value is: " + message.getMessage() + ".");
+        if (message.getLevel() == null) throw new IllegalArgumentException("Message message.getLevel() value is: " + message.getLevel() + ".");
+        return true;
```
Если метод валидации возвращает только true, а на любую не валидную ситуацию выкидывает исключение, то можно не возвращать true и сделать метод void

wrhmmstr replied:
Исправлено

Commit:
chore: изменения из репозитория лектора

dmitrii-grog commented on src/main/java/com/tcs/edu/service/DecoratingMessageService.java:
```diff
@@ -0,0 +1,161 @@
+     * @param messages  Массив строк (String...) с сообщениями для декорирования
+     * @see DecoratingMessageService Родительский класс
+     */
+    public Message[] combineMessages(Message message, Message... messages) {
```
Этот метод больше похож на внутренний, возможно клиентам его не нужно предоставлять, в таком случае метод должен быть private.

Но тогда этот метод неудобно тестировать, для этого логику комбинирования можно вынести в отдельный класс:
```
public class MessageCombiner {
    public Message[] combine(Message message, Message... messages) {
        ...
    }
}
```
В `DecoratingMessageService` его можно как передавать через конструктор, давая возможность клиентам предложить свою реализацию. а можно просто добавить как поле класса:
```
private final MessageCombiner messageCombiner = new MessageCombiner();
```
В результате логику комбинирования легко протестить в отдельном классе `MessageCombinerTest`

dmitrii-grog commented on src/main/java/com/tcs/edu/MessageProcessor.java:
```diff
@@ -0,0 +1,13 @@
+import com.tcs.edu.domain.MessageOrder;
+import com.tcs.edu.domain.Message;
+
+public interface MessageProcessor {
```
Идея хорошая - инкапсулировать логику сортировки и удаления дубликатов. 

Но опять же `MessageProcessor` и сортирует и удаляет дубликаты, класс отвечает за 2 действия сразу, лучше разделить на `OrderProcessor` и `DoublingProcessor` соответственно

И удобно для каждого обработчика свой класс тест

dmitrii-grog commented on src/main/java/com/tcs/edu/decorator/OrderDoublingProcessor.java:
```diff
@@ -0,0 +1,93 @@
+            case DISTINCT: {
+                int currentMessageIndex = 0;
+                int distinctWrittenMessageIndex = 0;
+                Message[] processedMessagesWithEmpties = new Message[messages.length];
```
Массивы это конечно хорошо, но реально работать с ними неудобно, поэтому если есть возможность используй подходящую коллекцию, например:
```suggestion
                final List<Message> distinctMessages = new ArrayList<>();
                for (Message message : messages) {
                    if (!distinctMessages.contains(message)) {
                        distinctMessages.add(message);
                    }
                }
                return distinctMessages.toArray(new Message[messages.length]);
```

dmitrii-grog commented on src/main/java/com/tcs/edu/service/DecoratingMessageService.java:
```diff
@@ -0,0 +1,161 @@
+     */
+    public void processMessages(MessageOrder order, Message message, Message... messages) throws LogException {
+        try {
+            processMessagesCycle(messageProcessor.process(order, combineMessages(message, messages)));
```
Возможно еще уменьшить дублирование, и вызывать перегруженные методы `processMessages ` по цепочке, подставляя дефолтные значения что не нарушат логику:

`processMessages(message, messages)` ->
`processMessages(MessageOrder.ASC, message, messages)` ->
`processMessages(Doubling.DOUBLES, MessageOrder.ASC, message, messages)`

в итоге конечная обработка будет только в одном методе:
```
        Message[] combinedMessages = messageCombiner.combine(message, messages);
        Message[] orderedMessages = orderProcessor.process(order, combinedMessages);
        Message[] processedMessages = doublingProcessor.process(doubling, orderedMessages);
        for (Message processedMessage : processedMessages) {
            processMessage(processedMessage);
        }
```
